server/etcdserver/api/etcdhttp/health.go

```diff
@@ -193,3 +224,135 @@ func checkAPI(lg *zap.Logger, srv ServerHealth, serializable bool) Health {
 	lg.Debug("serving /health true")
 	return h
 }
+
+type HealthCheck func(ctx context.Context) error
+
+type CheckRegistry struct {
+	path string
+
+	checks map[string]HealthCheck
+	// the list of checkNames are needed to ensure the order of checks.
+	checkNames []string
+}
+
+func installLivezEndpoints(lg *zap.Logger, mux *http.ServeMux, server ServerHealth) {
+	reg := CheckRegistry{path: "/livez", checks: make(map[string]HealthCheck)}
+	reg.Register("serializable_read", serializableReadCheck(server))
+	reg.InstallHttpEndpoints(lg, mux)
+}
+
+func installReayzEndpoints(lg *zap.Logger, mux *http.ServeMux, server ServerHealth) {
+	reg := CheckRegistry{path: "/readyz", checks: make(map[string]HealthCheck)}
+	reg.Register("data_corruption", activeAlarmCheck(server, pb.AlarmType_CORRUPT))
+	reg.Register("serializable_read", serializableReadCheck(server))
+	reg.InstallHttpEndpoints(lg, mux)
+}
+
+func (reg *CheckRegistry) Register(name string, check HealthCheck) {
+	reg.checkNames = append(reg.checkNames, name)
+	reg.checks[name] = check
+}
+
+func (reg *CheckRegistry) InstallHttpEndpoints(lg *zap.Logger, mux *http.ServeMux) {
+	// installs the http handler for the root path.
+	reg.installHttpEndpoint(lg, mux, reg.path, reg.checkNames...)
+	for _, checkName := range reg.checkNames {
+		// installs the http handler for the individual check sub path.
+		reg.installHttpEndpoint(lg, mux, path.Join(reg.path, checkName), checkName)
```

Comment:
Not sure I like that you are reusing `installHttpEndpoint` function to install checking of multiple and single checks. It unintentionally leaks `exclude` logic into the singular check logic. I have a strong preference for code to always be intentional and not have hidden side effects. After merging the code becomes the source of truth on how etcd should behave, obfuscating the intention.

I understand that this should not affect the calls if it's not used, but by adding to the API you make it a part of expected API behavior that adds unnecessary complexity. When someone will be refactoring this code in the future, they will need to understand if this is officially supported behavior. Some user could encounter it by mistake and start depending on this.

Reply:
done.